src/Namespace.sol

```diff
@@ -0,0 +1,578 @@
+            payable(msg.sender).transfer(msg.value - _currYearFee);
+        }
+    }
+
+    /**
+     * @notice Renew a name for another year while it is in the renewable period
+     *
+     * @param tokenId the tokenId of the name to renew
+     * @param owner the current owner of the name
+     */
+    function renew(uint256 tokenId, address owner) external payable {
+        if (msg.value < fee) revert InsufficientFunds();
+
+        uint256 expiryTs = expiryOf[tokenId];
+        if (expiryTs == 0) revert Registerable();
+
+        // Invariant 1B + 2 guarantee that the name is not owned by address(0) at this point.
+        if (_ownerOf[tokenId] != owner) revert IncorrectOwner();
+
+        unchecked {
+            // renewTs and gracePeriod are pre-determined values and cannot overflow
+            if (block.timestamp >= expiryTs + gracePeriod) revert Biddable();
+
+            if (block.timestamp < expiryTs) revert Registered();
+
+            // currYear is selected from a pre-determined list and cannot overflow
+            expiryOf[tokenId] = timestampOfYear(currYear() + 1);
+        }
+
+        emit Renew(tokenId, owner, expiryOf[tokenId]);
+
+        // TODO: this may fail if called by a smart contract
+        if (msg.value > fee) {
+            payable(msg.sender).transfer(msg.value - fee);
+        }
+    }
+
+    /**
+     * @notice Bid to purchase an expired username in the dutch auction, whose price is the sum of
+     *         the current year's fee and a premium. The premium is set to 1000 ether on Feb 1st
+     *         and decays by ~10% per period (8 hours) until it reaches zero mid-year.
+     *
+     * @dev The premium reduction is computed with the identity (x^y = exp(ln(x) * y)) with
+     *      gas-optimzied approximations for exp and ln that introduce a -3% error for every period
+     *
+     * @param tokenId the tokenId of the username to bid on
+     */
+    function bid(uint256 tokenId) external payable {
+        uint256 expiryTs = expiryOf[tokenId];
+        if (expiryTs == 0) revert Registerable();
+
+        uint256 auctionStartTimestamp;
+
+        unchecked {
+            // expiryTs is taken from a pre-determined list and cannot overflow.
+            auctionStartTimestamp = expiryTs + gracePeriod;
+        }
+
+        if (auctionStartTimestamp > block.timestamp) revert NotBiddable();
+
+        // Calculate the num of 8 hr periods since expiry as a fixed point signed decimal. The
+        // constant approximates fixed point division by 28,000 (num of seconds in 8 hours)
+        int256 periodsSD59x18 = int256(3.57142857e13 * (block.timestamp - auctionStartTimestamp));
+
+        // Optimization: precompute return values for the first few periods and the last one.
+
+        // Calculate the price by taking the 1000 ETH premium and discounting it by 10% for every
+        // period and adding to it the renewal fee for the current year.
+        uint256 price = uint256(1_000 ether).mulWadDown(
```

Comment:
Should `1_000 ether` be made a variable that the contract owner can update at a later time?

Otherwise, just for readability it could be made a named constant

Reply:
updateable values and named constants require slightly more gas, and the former also adds more surface area. i'd prefer to just do a full contract upgrade and keep the value immutable.